src/main/java/io/confluent/copycat/hdfs/TopicPartitionWriter.java

```diff
@@ -0,0 +1,461 @@
+    if (!m.matches()) {
+      return false;
+    }
+    String[] parts = filename.split("_");
+    String topic = parts[0];
+    int partition = Integer.parseInt(parts[1]);
+    return topic.equals(tp.topic()) && partition == tp.partition();
+  }
+
+  private void pause() {
+    context.pause(tp);
+  }
+
+  private void resume() {
+    context.resume(tp);
+  }
+
+  @SuppressWarnings("unchecked")
+  public RecordWriter<Long, SinkRecord> getWriter(SinkRecord record, String encodedPartition)
+      throws CopycatException {
+    try {
+      if (writers.containsKey(encodedPartition)) {
+        return writers.get(encodedPartition);
+      }
+      String tempfile = getTempfile(encodedPartition);
+      RecordWriter<Long, SinkRecord> writer = writerProvider.getRecordWriter(conf, tempfile, record, avroData);
+      writers.put(encodedPartition, writer);
+      return writer;
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private String getTempfile(String encodedPartition) {
+    String tempfile;
+    if (tempfiles.containsKey(encodedPartition)) {
+      tempfile = tempfiles.get(encodedPartition);
+    } else {
+      String directory = getDirectory(encodedPartition);
+      tempfile = FileUtils.tempFileName(url, topicsDir, directory);
+      tempfiles.put(encodedPartition, tempfile);
+    }
+    return tempfile;
+  }
+
+
+  private void createWAL() throws CopycatException {
+    if (wal == null) {
+      wal = storage.wal(topicsDir, tp);
+    }
+  }
+
+  private void applyWAL() throws CopycatException {
+    if (!recovered) {
+      wal.apply();
+    }
+  }
+
+  private void truncateWAL() throws CopycatException {
+    if (!recovered) {
+      wal.truncate();
+    }
+  }
+
+  private void resetOffsets() throws CopycatException {
+    if (!recovered) {
+      readOffset();
+      if (offset > 0) {
+        context.offset(tp, offset);
+      }
+      recovered = true;
+    }
+  }
+
+  private void writeRecord(SinkRecord record) throws IOException {
+    String encodedPartition = partitioner.encodePartition(record);
+    RecordWriter<Long, SinkRecord> writer = getWriter(record, encodedPartition);
+    writer.write(System.currentTimeMillis(), record);
+    if (offset == -1) {
+      offset = record.kafkaOffset() - 1;
+    }
+    if (!startOffsets.containsKey(encodedPartition)) {
+      startOffsets.put(encodedPartition, record.kafkaOffset() - 1);
+      offsets.put(encodedPartition, record.kafkaOffset() - 1);
+    } else {
+      offsets.put(encodedPartition, record.kafkaOffset() - 1);
+    }
+    recordCounter++;
+  }
+
+  private void closeTempfile(String encodedPartition) throws IOException {
+    if (writers.containsKey(encodedPartition)) {
+      RecordWriter writer = writers.get(encodedPartition);
+      writer.close();
+      writers.remove(encodedPartition);
+    }
+    recordCounter = 0;
+  }
+
+  private void closeTempfile() throws IOException {
+    for (String encodedPartition: tempfiles.keySet()) {
```

Comment:
I think I see the answer to my earlier question. Isn't this approach a problem if we write ? I think this patch basically makes it so we need atomic commits of multiple files since data can be spread across multiple files now. So we either need each entry in the WAL to include the list of renames to perform *or* we need some sort of start/end transaction records that indicate when all the information is available. Otherwise, `appendToWAL` might only manage to append a half the files and you'll deliver only half the data you expected to. As long as it's all guaranteed to be recorded in the WAL *or* we won't apply the half-completed write, it should be fine, but we need that guarantee.

By the way, this might be a good way to write a system test that stresses the WAL -- use a partitioning that causes a *lot* of files to be used in the same transaction. That'll drastically increase the likelihood of breaking the WAL/deliver guarantees.